### PR TITLE
Fix nested copy-and-update with complex original expressions

### LINKED ISSUE #16237: Compilation error from nested record field update
**Repro steps**

```fsharp
type Record1 = { Foo: int; Bar: int; }

[<AutoOpen>]
module Module =
    type Record2 = { Foo: Record1 }
    let item: Record2 = Unchecked.defaultof<Record2>

ignore { item        with Foo.Foo = 3 }
ignore { Module.item with Foo.Foo = 3 }
```

image

`{ Module.item with Foo.Foo = 3 }` should be compiled successfully, 
by analogy with  `{ item with Foo.Foo = 3 }`

**Related information**

* .NET 8.0.100-rc.2.23502.2

### PR DESCRIPTION
**Pull request by kerams**
Fixes #16237.

I'm not happy with this solution. I tried to put similar code into `TransformAstForNestedUpdates`, but the record checking functions create intermediary bindings too, so then I ended up with duplicates. This works, but feel free to refactor.

### CODE REVIEW

File: src/Compiler/Checking/CheckExpressions.fs [hunk 1]
```diff
@@ -5669,8 +5685,24 @@ and TcExprUndelayed (cenv: cenv) (overallTy: OverallTy) env tpenv (synExpr: SynE
...
+            TcExprRecord cenv overallTy env tpenv (inherits, withExprOpt, synRecdFields, mWholeExpr)
+        | Some(expr, blockSep) ->
+            let mOrigExprSynth = expr.Range.MakeSynthetic()
+            let id = mkSynId mOrigExprSynth "bind@"
+            let binding = mkSynBinding (PreXmlDoc.Empty, mkSynPatVar None id) (None, false, false, mOrigExprSynth, DebugPointAtBinding.NoneAtSticky, None, expr, mOrigExprSynth, [], [], None, SynBindingTrivia.Zero)
+
+            let withExpr = SynExpr.Ident id, blockSep
+
+            let body = SynExpr.Record (inherits, Some withExpr, synRecdFields, mWholeExpr)
+            let expr = SynExpr.LetOrUse (false, false, [ binding ], body, mOrigExprSynth, SynExprLetOrUseTrivia.Zero)
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckExpressions.fs:5703]: Can the shared body of the repeated block between AnonRecord and Record be extracted please?

**Review by T-Gro — APPROVED** [review]

File: src/Compiler/Checking/CheckRecordSyntaxHelpers.fsi [hunk 2]
```diff
@@ -19,3 +18,9 @@ val TransformAstForNestedUpdates<'a> :
     exprBeingAssigned: SynExpr ->
     withExpr: SynExpr * (range * 'a) ->
         (Ident list * Ident) * SynExpr option
+
+val BindOriginalRecdExpr:
+    withExpr: SynExpr * BlockSeparator ->
+        mkRecdExpr: ((SynExpr * BlockSeparator) option -> SynExpr) ->
+        tcExpr: (SynExpr -> Expr * UnscopedTyparEnv) ->
+            Expr * UnscopedTyparEnv
```
Comment (kerams) [inline src/Compiler/Checking/CheckRecordSyntaxHelpers.fsi:26]: Questionable formatting.

**Review by vzarytovskii — APPROVED** [review]
Nice, thanks

**Comment by edgarfgp** [comment]
This is not necessarily related to this PR but I have been chasing an compiler `internal error` when reporting duplicated fields in nested `anon records` and I have chased to this [line](https://github.com/dotnet/fsharp/blob/d441308a78724ea47584027af697789c8d2e1a94/src/Compiler/Checking/CheckExpressions.fs#L7638C9-L7638C9). Without doing `|> Array.distinctBy (fst >> textOfId)` the error reporting cascades properly. I was wondering if you happen to know we can avoid the  `Array.distinctBy (fst >> textOfId)` at that point to have a better error reporting for nested `anon records` .

**Review by edgarfgp — APPROVED** [review]

**Comment by auduchinok** [comment]
@kerams Do you know if there are any tests for reported symbols that cover these cases? It would be nice to assert that we still report usages of the symbols in the nested expressions correctly.

**Comment by kerams** [comment]
A couple have been added in https://github.com/dotnet/fsharp/pull/15699

**Pull request merged**